### PR TITLE
Add support for ARM shapes on Oracle OCI

### PR DESCRIPTION
**Pull request by nvinuesa**
By correctly parsing the version of the ubuntu cloud images returned by OCI's API, we can then create different bases for ubuntu aarch64 (the cloud image needed to target arm-based shapes).

It is important to note that we will not be supporting `Minimal aarch64` ubuntu images, based on what is documented for using ARM-based shapes: https://docs.oracle.com/en-us/iaas/Content/Compute/References/images.htm
> ...For Arm-based shapes, use the Ubuntu image, not Minimal Ubuntu.

Bonus: By moving the retrieve of `instanceTypes()` *after* we correctly parsed the `InstanceImage`, we avoid useless calls to the API to retrieve the shapes for images that we don't support anyways.
 
Bonus 2: Fixed the Makefile for building cross-compiled versions (thx to @hmlanigan and @tlm).

Bonus 3: Fixed the `TestRefreshImageCacheWithInvalidImage` test which had incorrect logic.

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

- Build
```
rm -rf _build/
make go-install
GOOS=linux GOARCH=arm64 make go-build
AGENT_PACKAGE_PLATFORMS=linux/arm64 make simplestreams
export JUJU_METADATA_SOURCE="/home/nicolas/workspace/canonical/juju/_build/simplestreams"
```
- Bootstrap and add model
```
juju bootstrap --debug --constraints arch=arm64 --config compartment-id=ocid1.compartment.oc1..aaaaaaaanvu2racnlczevenu73dlcf3nokgsjpdkbdgp4xbrz3lb2y4giyjq oci-canonical c
juju add-model m
```
- Add machines
```
juju add-machine --constraints "arch=arm64"
juju add-machine --constraints "arch=arm64 cores=3 mem=11G"
```

If you log in the OCI dashboard, you should see 3 `VM.Standard.A1.Flex` instances (one for the controller, two for the added machines) with 1 Ocpu for two of them and 3 Ocpus/ 11GB for the third one.

## Links

**Jira card:** [JUJU-4514](https://warthogs.atlassian.net/browse/JUJU-4514)

[JUJU-4514]: https://warthogs.atlassian.net/browse/JUJU-4514?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: provider/oci/environ.go [hunk 1]
```diff
@@ -498,20 +498,25 @@ func (e *Environ) startInstance(
...
 		args.Constraints.VirtType = &defaultType
 	}
 
+	// Since the images for ARM machine will include "aarch64" on their
+	// name, we need to make sure we select the correct base that was
+	// taken from the parsed image name.
+	if arch == corearch.ARM64 {
+		args.InstanceConfig.Base.OS = UbuntuARMBase
+	}
+
```
Comment (hmlanigan) [inline provider/oci/environ.go:512]: I have concerns about overloading the defined juju base.OS values with architecture details and the values ending up in state. Nor am I clear how this interacts with the architecture constraints.

What about updating imgCache.SupportedShapes to take an architecture value as well. The actual imgCache is local to the Oracle provider.

File: provider/oci/images.go [hunk 2]
```diff
@@ -381,37 +414,6 @@ func instTypeByShapeName(shape string) string {
...
-}{
-	{regexp.MustCompile(oracleAmd), arch.AMD64, VirtualMachine.String()},
-	{regexp.MustCompile(oracleAmdBm), arch.AMD64, BareMetal.String()},
-	{regexp.MustCompile(oracleAmdBmGpu), arch.AMD64, GPUMachine.String()},
-}
-
-// normaliseArchAndInstType returns the Juju architecture and instance type
-// corresponding to a shape's reported architecture and instance type based
-// off ShapePlatformConfigOptionsTypeEnum.
-func normaliseArchAndInstType(val ociCore.ShapePlatformConfigOptionsTypeEnum) (string, string) {
```
Comment (hmlanigan) [inline provider/oci/images.go:406]: You still need to handle the InstanceTypes here, VM, BM etc for use.

**Review by hmlanigan — CHANGES_REQUESTED** [review]
Images supporting Ubuntu 22.04 with amd64 are getting lost in the processing somewhere, not sure where: Image `ocid1.image.oc1.phx.aaaaaaaapj3tk2to46hrihgab6aytk3xfv5tcdbdms4k2ixzoz64zxlpmjfq` with String `Canonical-Ubuntu-22.04-aarch64-2023.08.23` is being parsed as `Canonical Ubuntu 22.04` per [MM yesterday](https://chat.charmhub.io/charmhub/pl/uhrgsxanjtgrpqkc6ybenj5hqw)

File: provider/oci/images.go [hunk 3]
```diff
@@ -251,31 +281,40 @@ func getImageType(img ociCore.Image) ImageType {
 	return ImageTypeGeneric
 }
 
-func NewInstanceImage(img ociCore.Image, compartmentID *string) (imgType InstanceImage, err error) {
+// NewInstanceImage returns a populated InstanceImage from the ociCore.Image
+// struct returned by oci's API.
+// Specifically, for ubuntu, we need to create a different base when the image
+// name contains "aarch64", which are the images that target ARM-based shapes.
+func NewInstanceImage(img ociCore.Image, compartmentID *string) (InstanceImage, error) {
+	logger.Warningf("*** CREATING INSTANCE IMAGE FROM '%s'//'%s'", *img.OperatingSystem, *img.OperatingSystemVersion)
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply provider/oci/images.go]: sure, I had forgotten to clean up after debugging, done

File: Makefile [hunk 4]
```diff
@@ -355,7 +355,7 @@ ${BUILD_DIR}/%/bin/pebble: phony_explicit
 # build for pebble
 	$(run_go_build)
 
-${JUJU_METADATA_SOURCE}/tools/released/juju-${JUJU_VERSION}-%.tgz: phony_explicit juju $(BUILD_AGENT_TARGETS) $(BUILD_CGO_AGENT_TARGETS)
+${JUJU_METADATA_SOURCE}/tools/released/juju-${JUJU_VERSION}-%.tgz: phony_explicit juju $(BUILD_AGENT_TARGETS)
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply Makefile]: I removed it for now, I'll add a new commit once we finish discussing the main items on this PR

File: provider/oci/images.go [hunk 5]
```diff
@@ -39,9 +37,22 @@ const (
 	centOS   = "CentOS"
 	ubuntuOS = "Canonical Ubuntu"
 
+	UbuntuBase        = "ubuntu"
+	UbuntuMinimalBase = "ubuntu-minimal"
+	UbuntuARMBase     = "ubuntu-" + arch.ARM64
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply provider/oci/images.go]: I changed the logic, this is gone now

File: provider/oci/images.go [hunk 6]
```diff
@@ -116,6 +127,21 @@ func (i *InstanceImage) SetInstanceTypes(types []instances.InstanceType) {
...
+func (i *InstanceImage) Archs() []string {
+	uniqueVals := make(map[string]bool)
+	var archs []string
+	for _, val := range i.InstanceTypes {
+		if !uniqueVals[val.Arch] {
+			archs = append(archs, val.Arch)
+		}
+		uniqueVals[val.Arch] = true
+	}
+	return archs
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply provider/oci/images.go]: oh nice! done

File: provider/oci/environ.go (hunk 1 above)
Comment (nvinuesa) [reply provider/oci/environ.go]: indeed, this was a very bad idea. I changed the implementation, now we pass the architecture as parameter to the image functions. Also, we keep track of the image's architecture on the cache so we can select them using the passed arch parameter.

File: provider/oci/images.go (hunk 2 above)
Comment (nvinuesa) [reply provider/oci/images.go:406]: this is done in `instTypeByShapeName()` which I didn't touch

**Comment by hmlanigan** [comment]
@nvinuesa once a review has started, please add new commits with changes rather than force pushing an amended commit (excepting rebases). It makes it more difficult for reviewers. They can be squashed before merge.

**Comment by hmlanigan** [comment]
This PR is blocked while @nvinuesa investigates improved image and instance modeling for juju where Flexible instances exist in a cloud.

File: provider/oci/images.go [hunk 7]
```diff
@@ -331,12 +370,6 @@ func instanceTypes(cli ComputeClient, compartmentID, imageID *string) ([]instanc
 }
 
 func parseArchAndInstType(shape ociCore.Shape) (string, string) {
-	// This code is very brittle. It's highly dependent on the strings currently
-	// returned by the oracle sdk. The best option is to have
-	// PlatformConfigOptions as they have types we can rely on.
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply provider/oci/images.go:353]: but we are no longer using `PlatformConfigOptions`, what comment should we keep from that?

**Comment by hmlanigan** [comment]
@nvinuesa with the qa instructions, I fail to bootstrap with:
```
18:36:11 ERROR juju.cmd.juju.commands bootstrap.go:988 failed to bootstrap model: cannot use agent built for "arm64" using a machine running on "amd64"
18:36:11 DEBUG juju.cmd.juju.commands bootstrap.go:989 (error details: [{github.com/juju/juju/cmd/juju/commands.(*bootstrapCommand).Run:1087: failed to bootstrap model} {github.com/juju/juju/environs/bootstrap.Bootstrap:732: } {cannot use agent built for "arm64" using a machine running on "amd64"}])
```

File: provider/oci/images.go [hunk 8]
```diff
@@ -39,6 +40,8 @@ const (
 	centOS   = "CentOS"
 	ubuntuOS = "Canonical Ubuntu"
 
+	UbuntuBase = "ubuntu"
```
Comment (hmlanigan) [inline provider/oci/images.go:43]: isn't this a const somewhere already?

File: provider/oci/images.go [hunk 9]
```diff
@@ -140,15 +152,16 @@ func (t byVersion) Less(i, j int) bool {
 	return true
 }
 
+type imageMap map[corebase.Base]map[string][]InstanceImage
```
Comment (hmlanigan) [inline provider/oci/images.go:155]: Given this is a complex type, please comment on what each value holds, "string" could be anything.

File: provider/oci/images.go [hunk 10]
```diff
@@ -190,27 +203,31 @@ func (i ImageCache) ImageMetadata(base corebase.Base, defaultVirtType string) []
...
-		imgMeta := &imagemetadata.ImageMetadata{
-			Id:   val.Id,
-			Arch: "amd64",
-			// TODO(gsamfira): add region name?
-			// RegionName: region,
-			Version:  val.Base.Channel.Track,
-			VirtType: string(val.ImageType),
+		// Iterate over the list of supported architectures of the
+		// image.
+		for _, arch := range val.Archs() {
```
Comment (hmlanigan) [inline provider/oci/images.go:208]: Why do we need to iterate over val.Arches(), if val can only be 1 arch?

File: provider/oci/images.go [hunk 11]
```diff
@@ -251,31 +268,30 @@ func getImageType(img ociCore.Image) ImageType {
 	return ImageTypeGeneric
 }
 
-func NewInstanceImage(img ociCore.Image, compartmentID *string) (imgType InstanceImage, err error) {
-	var base corebase.Base
+// NewInstanceImage returns a populated InstanceImage from the ociCore.Image
+// struct returned by oci's API. Also, it returns the architecture that is
+// detected from the image OperatingSystemVersion and it's name.
```
Comment (hmlanigan) [inline provider/oci/images.go:273]: The 2nd sentence doesn't match the function return values.

s/it's/its/

File: provider/oci/images.go [hunk 12]
```diff
@@ -284,11 +300,56 @@ func NewInstanceImage(img ociCore.Image, compartmentID *string) (imgType Instanc
...
+// from the OCI sdk.
+func parseUbuntuImage(img ociCore.Image) (corebase.Base, string, error) {
+	var (
+		arch string
+		base corebase.Base
+	)
+	// we need to split the provided OperatingSystemVersion, since
+	// it can contain information that does not belong to the base's
+	// channel.
+	// E.g: Canonical-Ubuntu-22.04-Minimal-aarch64-2023.08.27-0
```
Comment (hmlanigan) [inline provider/oci/images.go:320]: More examples of possible OperatingSystemVersion values would be helpful.

File: provider/oci/images.go [hunk 13]
```diff
@@ -284,11 +300,56 @@ func NewInstanceImage(img ociCore.Image, compartmentID *string) (imgType Instanc
...
+	// becomes:
+	//        OperatingSystem:        Canonical Ubuntu
+	//        OperatingSystemVersion: 22.04 Minimal aarch64
```
Comment (hmlanigan) [inline provider/oci/images.go:323]: This was copied from older code, and should not be valid with your fixes. Please validate.

**Review by hmlanigan — COMMENTED** [review]
I need more time to review. The parsing is complex. The qa did not work as written.

File: provider/oci/images.go (hunk 8 above)
Comment (nvinuesa) [reply provider/oci/images.go:43]: there might be but I haven't found one, I tried searching many different regexes and nothing came up. We have ~17000 occurrences of "ubuntu" so it might be there somewhere :man_shrugging:

File: provider/oci/images.go (hunk 10 above)
Comment (nvinuesa) [reply provider/oci/images.go:208]: the issue is that we cannot enforce that. Since the `instanceTypes` are created from the returned list of shapes from the API, we can't be sure that it will always be the case. You suggest that we throw an error if `len(val.Archs()) > 1`?

File: provider/oci/images.go (hunk 9 above)
Comment (nvinuesa) [reply provider/oci/images.go:157]: you're absolutely right. Let me know if the description I added looks good or you would rephrase it

File: provider/oci/images.go (hunk 11 above)
Comment (nvinuesa) [reply provider/oci/images.go]: I just rephrased it

File: provider/oci/images.go (hunk 12 above)
Comment (nvinuesa) [reply provider/oci/images.go]: I added a second example, with more detailed explanation

File: provider/oci/images.go (hunk 13 above)
Comment (nvinuesa) [reply provider/oci/images.go]: I rephrased this comment

File: provider/oci/images.go (hunk 8 above)
Comment (hmlanigan) [reply provider/oci/images.go:42]: I eventually found one, but it's defined with "Ubuntu". Juju doesn't care about capitalization, but some pieces do, let's leave this.

**Review by hmlanigan — APPROVED** [review]
This one took a while. It was worth it to improve how juju models instance selections for newer types of cloud machines.
